src/validation.cpp

```diff
@@ -1608,16 +1608,28 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
 // Protected by cs_main
 static ThresholdConditionCache warningcache[VERSIONBITS_NUM_BITS];
 
+static bool IsScriptWitnessEnabled(const Consensus::Params& params)
+{
+    return params.vDeployments[Consensus::DEPLOYMENT_SEGWIT].nTimeout != 0;
+}
+
 static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consensus::Params& consensusparams) {
     AssertLockHeld(cs_main);
 
     unsigned int flags = SCRIPT_VERIFY_NONE;
 
-    // Start enforcing P2SH (BIP16)
-    if (pindex->nHeight >= consensusparams.BIP16Height) {
+    // BIP16 didn't become active until Apr 1 2012
+    // However, only one historical block violated the P2SH rules, so for simplicity, always leave P2SH
+    // on except for the one violating block.
+    if (!(pindex->phashBlock != nullptr && (*(pindex->phashBlock) == consensusparams.BIP16Exception))) {
```

Comment:
Why not assert pindex->phashBlock? No properly-constructed CBlockIndex should have that unset, no?

Reply:
It's unset when invoked via TestBlockValidity() (eg for mining).

Reply:
Ah, that sounds like a bug in TBV, but that's for a separate PR, I suppose.

Reply:
Yes, please, I always found checks to pindex->phashBlock == nullptr ugly and expensive to maintain documented (perhaps we can repeat the "pindex->phashBlock can be null if called by CreateNewBlock/TestBlockValidity" comment from the beginning and unified with the comment for fEnforceBIP30, to more easily delete them all later when they're not needed). There's the asset you want below, just needs to be moved to the top once the special cases are dealt with more explicitly.
But that's for a separate PR.